src/go/k8s/pkg/resources/certmanager/pki.go

```diff
@@ -147,3 +143,138 @@ func (r *PkiReconciler) Ensure(ctx context.Context) error {
 func (r *PkiReconciler) issuerNamespacedName(name string) types.NamespacedName {
 	return types.NamespacedName{Name: r.pandaCluster.Name + "-" + name, Namespace: r.pandaCluster.Namespace}
 }
+
+func (r *PkiReconciler) prepareAPI(
+	ctx context.Context,
+	rootCertSuffix string,
+	nodeCertSuffix string,
+	clientCerts []string,
+	listeners []APIListener,
+	keystoreSecret *types.NamespacedName,
+) ([]resources.Resource, error) {
+	var (
+		tlsListener         = getTLSListener(listeners)
+		toApply             = []resources.Resource{}
+		externalTLSListener = getExternalTLSListener(listeners)
+		internalTLSListener = getInternalTLSListener(listeners)
+		nodeSecretRef       *corev1.ObjectReference
+		// Default issuer for all certs
+		issuerRef *cmmetav1.ObjectReference
+		// Issuer for the nodes
+		nodeIssuerRef *cmmetav1.ObjectReference
+	)
+
+	if tlsListener == nil {
+		return []resources.Resource{}, nil
+	}
+
+	toApplyRoot, issuerRef := r.prepareRoot(rootCertSuffix)
+	toApply = append(toApply, toApplyRoot...)
+	nodeIssuerRef = issuerRef
+
+	if tlsListener != nil && tlsListener.GetTLS() != nil {
+		if tlsListener.GetTLS().IssuerRef != nil {
+			// if external issuer is provided, we will use it to generate node certificates
+			nodeIssuerRef = tlsListener.GetTLS().IssuerRef
+		}
+		nodeSecretRef = tlsListener.GetTLS().NodeSecretRef
+	}
+
+	if nodeSecretRef == nil {
+		certName := NewCertName(r.pandaCluster.Name, nodeCertSuffix)
+		certsKey := types.NamespacedName{Name: string(certName), Namespace: r.pandaCluster.Namespace}
+		dnsNames := []string{}
+
+		if internalTLSListener != nil {
+			dnsNames = append(dnsNames, r.internalFQDN)
+		}
+		// TODO: Add support for external listener + TLS certs for IPs
+		if externalTLSListener != nil && externalTLSListener.GetExternal().Subdomain != "" {
+			dnsNames = append(dnsNames, externalTLSListener.GetExternal().Subdomain)
+		}
+
+		nodeCert := NewNodeCertificate(r.Client, r.scheme, r.pandaCluster, certsKey, nodeIssuerRef, dnsNames, EmptyCommonName, keystoreSecret, r.logger)
+
+		toApply = append(toApply, nodeCert)
+	}
+
+	if nodeSecretRef != nil && nodeSecretRef.Namespace != r.pandaCluster.Namespace {
+		if err := r.copyNodeSecretToLocalNamespace(ctx, nodeSecretRef); err != nil {
+			return nil, err
```

Comment:
I found that function `prepareAPI` can return error only in this place. The error will be hard to find because while the error bubbles up the stack none of the function adds context to it. Places to consider add context to the error:

* https://github.com/vectorizedio/redpanda/blob/7877db314af4002b06d3ce03d9f8feebd23e0b89/src/go/k8s/pkg/resources/certmanager/pki.go#L104
* https://github.com/vectorizedio/redpanda/blob/7877db314af4002b06d3ce03d9f8feebd23e0b89/src/go/k8s/pkg/resources/certmanager/pki.go#L112
* https://github.com/vectorizedio/redpanda/blob/7877db314af4002b06d3ce03d9f8feebd23e0b89/src/go/k8s/pkg/resources/certmanager/pki.go#L120
* https://github.com/vectorizedio/redpanda/blob/7877db314af4002b06d3ce03d9f8feebd23e0b89/src/go/k8s/pkg/resources/certmanager/pki.go#L128
* https://github.com/vectorizedio/redpanda/blob/7877db314af4002b06d3ce03d9f8feebd23e0b89/src/go/k8s/pkg/resources/certmanager/pki.go#L203

I would vote to create here generic context like:
```suggestion
			return nil, fmt.Errorf("copy node secret for %s cert group to namespace %s: %w", rootCertSuffix, nodeSecretRef.Namespace, err)
```

Reply:
✅